Create BombingUnitDamageChange that uses primitives

trevan opened the pull request:
The old BombingUnitDamageChange is marked as deprecated and will stay
around solely for loading old saves.

The new BombingUnitDamageChange also has been modified to accept the
territory that needs to be notified, instead of looping through all the
game territories during the change process.


<!--
  Commit comment above summarizing the update.  If multiple commits please
  summarize the change above.
  Code standards and PR guidelines can be found at:
  <https://github.com/triplea-game/triplea/wiki/Contribution-Guidelines>
-->

## Testing
<!-- Describe any manual testing performed below. -->
Conducted a strategic bombing run in WW2 Global 1940.  Verified the change was correctly built.  Verified that it "performed" correctly.  Played till the attacked player could repair the damaged units and verified that the change was correctly built.
## Screens Shots
<!-- If there are UI updates, include screenshots below -->

## Additional Notes to Reviewer
<!-- Add any additional details that would be helpful to reviewers -->
I created a new Change object and marked the other as deprecated.  I think this might be cleaner since the new change object can have a new name and it doesn't need to hold the non-primitive objects.  This definitely breaks forward compatibility but the compatibility check should prevent users from trying to load saves with this new change object in an older version.

## Release Note

<!--
Include a release note if there is a bug fix or a visible change for players.
For format & syntax help, see:
https://github.com/triplea-game/triplea/blob/master/docs/pr-release-notes.md
-->

<!--RELEASE_NOTE--><!--END_RELEASE_NOTE-->

trevan commented:
The only difference between the old BombingUnitDamageChange class and the new one is the primitive usage.

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/data/changefactory/units/BombingUnitDamageChange.java:
```diff
@@ -0,0 +1,71 @@
+public class BombingUnitDamageChange extends Change {
+  private static final long serialVersionUID = 7093658184880237574L;
+  private final IntegerMap<String> hits;
+  private final IntegerMap<String> undoHits;
```
What is undo hits representing?

trevan replied:
undoHits represents the inverse of what is happening in this change.  Each change object has an "undo" version of it.

trevan replied:
The change objects call it "invert".  You can see it being used in the `invert` method.

DanVanAtta replied:
What would you think to rename:
 'hits' -> 'newHits' | 'hitsAfterChange'
'undoHits' -> 'oldHits' | 'hitsBeforeChange' | 'originalHits'

It looks like other change objects use 'new' & 'old' for naming. The 'undoHits' is not the best name, not really clear whether it is the original hits or a delta to compute the original hits.

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/data/changefactory/units/BombingUnitDamageChange.java:
```diff
@@ -0,0 +1,71 @@
+  private final IntegerMap<String> undoHits;
+  private final Collection<String> territoriesToNotify;
+
+  private BombingUnitDamageChange(
```
Any thoughts to use an `@AllArgsConstructor(visibility = AccessLevel.PRIVATE)` here to avoid the manually written out constructor?

trevan replied:
I could.  This is just copied code so I didn't make any major changes except the non-primitive -> primitive change.

DanVanAtta replied:
Moving code is one thing, in this case we are re-writing and in my 2 cents treat this is as new code.

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/data/changefactory/units/BombingUnitDamageChange.java:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  @Override
+  public String toString() {
```
Is the 'toString' required? Is it used anywhere to print text for users? If not, is it helpful?

trevan replied:
No idea.  Several change objects have a toString.  Several don't.

DanVanAtta replied:
A lot of 'toStrings' are included to help debugging in Eclipse that used to show the 'toString' of an object when in debug. That was before when you could inspect the values of an object and you'd see just the 'toString' value.

If you play a game with a break-point in this method you'll probably be able to determine if this is dead code relatively quickly. IMO we should not be porting dead-code into new code if we can help it. When we do a new release I don't think we'll want to go back and do any clean up beyond carefully deleting that which is marked for deletion.

trevan commented:
@DanVanAtta , I used the `@AllArgsConstructor` and answered your other questions.

trevan commented:
The build failed because of a failure in downloading a file.

trevan commented:
@DanVanAtta , I saw this as just a code move since the only reason the old file still exists is because of save compatibility.  But if you want to see it as re-writing, that's fine.  I've renamed the variables to `oldDamage` and `newDamage`, removed the `toString` (the debugger never stopped in it), and added some comments to explain what the value of `damage` should be.

DanVanAtta commented:
Thanks for the updates @trevan .  I'm pretty positive when we remove the old code that we're not going to then improve this moved and slightly changed version, which implies to me now is the best time to make it as reasonable as we can (just as would generally be asked for anything that is added).
